src/main/java/com/coveo/pushapiclient/Sandbox.java

```diff
@@ -0,0 +1,88 @@
+package com.coveo.pushapiclient;
+
+import java.util.HashMap;
+import java.util.List;
+
+public class Sandbox {
+
+    /**
+     * Upload a document batch in push mode -> using the /files endpoint
+     */
+    public static void uploadBatch() {
+        CatalogSource source = new CatalogSource("my_api_key", "my_org_id");
+
+        // Prepare the push in "Update" mode. In this mode, the push Service will create
+        // S3 file containers with appropriate batches sizes.
+        DocumentUpdateService service = source.startDocumentUpdate("my_source_id");
+
+        // Create a stream for pushing documents
+
+        // Prepare a list of documents to add or update from an imaginary method
+        // The SDK handles the batching. Simply feed documents into the service
+        for (DocumentBuilder document : prepareDocuments()) {
+            service.addOrUpdateDocument(document);
+        }
+
+        // Prepare a list of documents to partially update from an imaginary method
+        // Same method can be used for both full and partial updates
+        for (PartialUpdateDocument document : preparePartialUpdateDocuments()) {
+            service.addOrUpdateDocument(document);
+        }
+
+        // Prepare a list of documents to delete from an imaginary method
+        for (DeleteDocument document : prepareDocumentsToDelete()) {
+            service.deleteDocument(document);
+        }
+
+        // Flush any previous documents buffered and not yet sent to the API.
+        service.flush();
+    }
+
+    /**
+     * Upload a document batch in stream mode -> using the /chunk endpoint
+     */
+    public static void streamBatch() {
+        CatalogSource source = new CatalogSource("my_api_key", "my_org_id");
+
+        // Prepare the push in "Stream" mode. In this mode, the push Service opens a
+        // stream and creates the appropriates stream chunks
+        FullCatalogUploadService service = source.startfullCatalogUpload("my_source_id");
+
+        // Prepare a list of documents to add from an imaginary method
+        // The SDK handles the batching.
+        for (DocumentBuilder document : prepareDocuments()) {
+            service.addDocument(document);
+        }
+        // Flush any previous documents buffered and not yet sent to the API.
+        // Closes the stream
+        service.flush();
```

Comment:
It's not clear who is calling this? Is it the sdk user, or internal?
And should `close` should be coupled to it? One might want to "flush" multiple times before closing the stream?

Reply:
The user is calling `flush`. 
If the Stream mode was chosen, then yes, the stream will be closed.

Even, if the user flushes multiple time, the SDK will still ensure the documents are properly sent.

Reply:
yep but if we close the stream then we send more documents after , it will open another stream ? so it will override what was already sent ? my understanding is that between open the stream and close it , all the catalog products should be sent

Reply:
Wondering if we could hide this from the implementer, by checking whether or not to flush on behalf of the user every time the `addOrUpdate`, `delete` or `close` methods are called? Putting this responsibility on the implementer means they need to ensure that they're within the [256 MB stream chunk limit](https://docs.coveo.com/en/lb4a0344/coveo-for-commerce/stream-your-catalog-data-to-your-source#:~:text=When%20the%20payload%20exceeds%20256%20MB%2C%20it%20must%20be%20chunked%20in%20256%20MB%20parts.).

I think this might trip people, and as a reaction, cause them to flush after every `addOrUpdate`, creating a lot of chunks which does not feel ideal for the platform (I am speculating on the last point though). Even if the platform can handle it though, hiding it feels like a simpler api,

Reply:
> it will open another stream ?
No. It will close the stream 
> so it will override what was already sent?
No, it will push buffered documents and close the stream. All documents added by the user before flushing will be sent.

Maybe the word `flush` is not ideal.
The goal of `flushing` is to actually take the batching responsibility away from the user. The idea is to actually let the users push their documents without caring about the chunk limit. The SDK takes care of the batching. However, the SDK needs to know when the user has no longer documents to push so the ones that were buffered (not yet added sent in a chunk) can be sent.

We could rename `flush` with `close` if you think it's less confusing.

Reply:
>  but an implementer could choose to call .close() explicitly?

We could make the choice to make it available yeah

> With a try/catch, if a genuine error happens (e.g. during prepareDocuments), implicitly closing the stream will cause a partial catalog upload. If the error occurs when ~50% of documents have been added, the catalog will be missing half the products? So .close should really be called only after a successful for loop versus always as part of a "finally" block. It feels like the try pattern might be misleading?

Mhh, I'm unsure what is the lesser evil here: partial update, or leaving the stream open?

Reply:
So the `autoclose` will not actually close the opened stream. It will only push buffered documents (when possible), which should **always execute**.

The class will then have a `close` method that will definitively have to call to close the stream (relinquishing any underlying resources).
```java
StreamSource source = new StreamSource("source_url", "my_api_key");
try (FullUploadStream stream = new FullUploadStream(source)) {
    // Prepare a list of documents to add from an imaginary method
    // The SDK handles the batching.
    for (DocumentBuilder document : prepareDocuments()) {
        stream.add(document);
    }
    stream.close()
}
```

Reply:
> It will only push buffered documents (when possible), which should always execute.

Hmm, so is `FullUploadStream` still implementing the `Autoclosable` interface? Or a different class that's used under-the-hood? It feels like the push can be done by `add` under-the-hood if the queue meets some criteria? e.g.

```
function add(Document doc) {
   queue.add(doc)
   
   if (queue.size > 256 MB) {
     chunk = extractChunk(queue)
     push(chunk)
   }
}
```

Reply:
@samisayegh , 
yes, that's the idea... but how can you push remaining documents in the queue when the size has not exceeded 256MB?
The idea behind the `AutoCloseable` interface is to push that remaining queue.

Reply:
yea, the sdk could push anything left in the queue when calling `.close()`, followed by closing the stream?